### ISSUE TITLE
`answer_component` updates

### ISSUE THREAD

**Issue by vishalbakshi** [vishalbakshi/fastbook-benchmark#1]
During my manual evaluation of MRR@10 and Recall@10 for a 500-token chunk size using BM25, single-vector, ColBERTv2 and answerai-colbert-small-v1, I thought of a few updates to this dataset, with the goal to remove what I think is unnecessary context. With this context removed, one or more methods would have retrieved the relevant passage(s):

---

- [ ] Shorten Ch1 Q30 context - contains extraneous segmentation definition info

Current:

> For instance, let's talk about something that is critically important for autonomous vehicles: localizing objects in a picture. If a self-driving car doesn't know where a pedestrian is, then it doesn't know how to avoid one! Creating a model that can recognize the content of every individual pixel in an image is called *segmentation*. Here is how we can train a segmentation model with fastai, using a subset of the [*Camvid* dataset](http://www0.cs.ucl.ac.uk/staff/G.Brostow/papers/Brostow_2009-PRL.pdf) from the paper \"Semantic Object Classes in Video: A High-Definition Ground Truth Database\" by Gabruel J. Brostow, Julien Fauqueur, and Roberto Cipolla:

Proposed:
> Creating a model that can recognize the content of every individual pixel in an image is called *segmentation*

---

- [ ] Shorten Ch2 Q6 context - remove "Jeremy likes..."

Current:
> However, nearly all machine learning approaches have the downside that they only tell you what products a particular user might like, rather than what recommendations would be helpful for a user. Many kinds of recommendations for products a user might like may not be at all helpful\u2014for instance, if the user is already familiar with the products, or if they are simply different packagings of products they have already purchased (such as a boxed set of novels, when they already have each of the items in that set). Jeremy likes reading books by Terry Pratchett, and for a while Amazon was recommending nothing but Terry Pratchett books to him (see <<pratchett>>), which really wasn't helpful because he already was aware of these books!

Proposed:
> However, nearly all machine learning approaches have the downside that they only tell you what products a particular user might like, rather than what recommendations would be helpful for a user. Many kinds of recommendations for products a user might like may not be at all helpful\u2014for instance, if the user is already familiar with the products, or if they are simply different packagings of products they have already purchased (such as a boxed set of novels, when they already have each of the items in that set)

---

- [ ] Remove second answer component of Ch4 Q14 and put its context as option for first answer component

Current:

```
{
            "chapter": 4,
            "question_number": 14,
            "question_text": "\"What are the seven steps in SGD for machine learning?\"",
            "gold_standard_answer": "\"Initialize the parameters \u2013 Random values often work best.\nCalculate the predictions \u2013 This is done on the training set, one mini-batch at a time.\nCalculate the loss \u2013 The average loss over the minibatch is calculated\nCalculate the gradients \u2013 this is an approximation of how the parameters need to change in order to minimize the loss function\nStep the weights \u2013 update the parameters based on the calculated weights\nRepeat the process\nStop \u2013 In practice, this is either based on time constraints or usually based on when the training/validation losses and metrics stop improving.\"",
            "answer_context": [
                {
                    "answer_component": "\"Initialize the parameters \u2013 Random values often work best.\nCalculate the predictions \u2013 This is done on the training set, one mini-batch at a time.\nCalculate the loss \u2013 The average loss over the minibatch is calculated\nCalculate the gradients \u2013 this is an approximation of how the parameters need to change in order to minimize the loss function\nStep the weights \u2013 update the parameters based on the calculated weights\nRepeat the process\nStop \u2013 In practice, this is either based on time constraints or usually based on when the training/validation losses and metrics stop improving.\"", 
                    "scoring_type": "simple",
                    "context": [
                        "1. *Initialize* the weights.\n1. For each image, use these weights to *predict* whether it appears to be a 3 or a 7.\n1. Based on these predictions, calculate how good the model is (its *loss*).\n1. Calculate the *gradient*, which measures for each weight, how changing that weight would change the loss\n1. *Step* (that is, change) all the weights based on that calculation.\n1. Go back to the step 2, and *repeat* the process.\n1. Iterate until you decide to *stop* the training process (for instance, because the model is good enough or you don't want to wait any longer)."
                    ],
                    "explicit_context": "true",
                    "extraneous_answer": "false"
                },
                {
                    "answer_component": "\"Initialize the parameters \u2013 Random values often work best.\nCalculate the predictions \u2013 This is done on the training set, one mini-batch at a time.\nCalculate the loss \u2013 The average loss over the minibatch is calculated\nCalculate the gradients \u2013 this is an approximation of how the parameters need to change in order to minimize the loss function\nStep the weights \u2013 update the parameters based on the calculated weights\nRepeat the process\nStop \u2013 In practice, this is either based on time constraints or usually based on when the training/validation losses and metrics stop improving.\"", 
                    "scoring_type": "simple",
                    "context": [
                        "- Initialize:: We initialize the parameters to random values. This may sound surprising. There are certainly other choices we could make, such as initializing them to the percentage of times that pixel is activated for that category\u2014but since we already know that we have a routine to improve these weights, it turns out that just starting with random weights works perfectly well.\n- Loss:: This is what Samuel referred to when he spoke of *testing the effectiveness of any current weight assignment in terms of actual performance*. We need some function that will return a number that is small if the performance of the model is good (the standard approach is to treat a small loss as good, and a large loss as bad, although this is just a convention).\n- Step:: A simple way to figure out whether a weight should be increased a bit, or decreased a bit, would be just to try it: increase the weight by a small amount, and see if the loss goes up or down. Once you find the correct direction, you could then change that amount by a bit more, and a bit less, until you find an amount that works well. However, this is slow! As we will see, the magic of calculus allows us to directly figure out in which direction, and by roughly how much, to change each weight, without having to try all these small changes. The way to do this is by calculating *gradients*. This is just a performance optimization, we would get exactly the same results by using the slower manual process as well.\n- Stop:: Once we've decided how many epochs to train the model for (a few suggestions for this were given in the earlier list), we apply that decision. This is where that decision is applied. For our digit classifier, we would keep training until the accuracy of the model started getting worse, or we ran out of time."
                    ],
                    "explicit_context": "true",
                    "extraneous_answer": "false"
                }
            ],
            "question_context": []
        }
```

Proposed:

```
{
            "chapter": 4,
            "question_number": 14,
            "question_text": "\"What are the seven steps in SGD for machine learning?\"",
            "gold_standard_answer": "\"Initialize the parameters \u2013 Random values often work best.\nCalculate the predictions \u2013 This is done on the training set, one mini-batch at a time.\nCalculate the loss \u2013 The average loss over the minibatch is calculated\nCalculate the gradients \u2013 this is an approximation of how the parameters need to change in order to minimize the loss function\nStep the weights \u2013 update the parameters based on the calculated weights\nRepeat the process\nStop \u2013 In practice, this is either based on time constraints or usually based on when the training/validation losses and metrics stop improving.\"",
            "answer_context": [
                {
                    "answer_component": "\"Initialize the parameters \u2013 Random values often work best.\nCalculate the predictions \u2013 This is done on the training set, one mini-batch at a time.\nCalculate the loss \u2013 The average loss over the minibatch is calculated\nCalculate the gradients \u2013 this is an approximation of how the parameters need to change in order to minimize the loss function\nStep the weights \u2013 update the parameters based on the calculated weights\nRepeat the process\nStop \u2013 In practice, this is either based on time constraints or usually based on when the training/validation losses and metrics stop improving.\"", 
                    "scoring_type": "simple",
                    "context": [
                        "1. *Initialize* the weights.\n1. For each image, use these weights to *predict* whether it appears to be a 3 or a 7.\n1. Based on these predictions, calculate how good the model is (its *loss*).\n1. Calculate the *gradient*, which measures for each weight, how changing that weight would change the loss\n1. *Step* (that is, change) all the weights based on that calculation.\n1. Go back to the step 2, and *repeat* the process.\n1. Iterate until you decide to *stop* the training process (for instance, because the model is good enough or you don't want to wait any longer).",

                       "- Initialize:: We initialize the parameters to random values. This may sound surprising. There are certainly other choices we could make, such as initializing them to the percentage of times that pixel is activated for that category\u2014but since we already know that we have a routine to improve these weights, it turns out that just starting with random weights works perfectly well.\n- Loss:: This is what Samuel referred to when he spoke of *testing the effectiveness of any current weight assignment in terms of actual performance*. We need some function that will return a number that is small if the performance of the model is good (the standard approach is to treat a small loss as good, and a large loss as bad, although this is just a convention).\n- Step:: A simple way to figure out whether a weight should be increased a bit, or decreased a bit, would be just to try it: increase the weight by a small amount, and see if the loss goes up or down. Once you find the correct direction, you could then change that amount by a bit more, and a bit less, until you find an amount that works well. However, this is slow! As we will see, the magic of calculus allows us to directly figure out in which direction, and by roughly how much, to change each weight, without having to try all these small changes. The way to do this is by calculating *gradients*. This is just a performance optimization, we would get exactly the same results by using the slower manual process as well.\n- Stop:: Once we've decided how many epochs to train the model for (a few suggestions for this were given in the earlier list), we apply that decision. This is where that decision is applied. For our digit classifier, we would keep training until the accuracy of the model started getting worse, or we ran out of time."
                    ],
                    "explicit_context": "true",
                    "extraneous_answer": "false"
                }
            ],
            "question_context": []
        }
```

- [ ] Remove "\u2014" from Ch4 Q16 contexts in benchmark

Current:
> The purpose of the loss function is to measure the difference between predicted values and the true values \\u2014 that is, the targets (aka labels)

Proposed:
> The purpose of the loss function is to measure the difference between predicted values and the true values — that is, the targets (aka labels)

---

- [ ] Shorten Ch4 Q19 context - remove "What's more..."

Current:
> We mentioned just now that you won't have to calculate any gradients yourself. How can that be? Amazingly enough, PyTorch is able to automatically compute the derivative of nearly any function! What's more, it does it very fast. Most of the time, it will be at least as fast as any derivative function that you can create by hand. Let's see an example.

Proposed:
> We mentioned just now that you won't have to calculate any gradients yourself. How can that be? Amazingly enough, PyTorch is able to automatically compute the derivative of nearly any function

---

- [ ] Shorten Ch8 Q27 context - remove "We will see in the next section..."

Current:
> Better still is to use a tabular model based on user meta data to construct your initial embedding vector. When a user signs up, think about what questions you could ask them that could help you to understand their tastes. Then you can create a model where the dependent variable is a user's embedding vector, and the independent variables are the results of the questions that you ask them, along with their signup metadata. We will see in the next section how to create these kinds of tabular models. (You may have noticed that when you sign up for services such as Pandora and Netflix, they tend to ask you a few questions about what genres of movie or music you like; this is how they come up with your initial collaborative filtering recommendations.)

Proposed:
> Better still is to use a tabular model based on user meta data to construct your initial embedding vector. When a user signs up, think about what questions you could ask them that could help you to understand their tastes. Then you can create a model where the dependent variable is a user's embedding vector, and the independent variables are the results of the questions that you ask them, along with their signup metadata

---

- [ ] Shorten Ch8 Q29 context - remove "fastai has a function..."

Current:
> Since we'll be concatenating the embeddings, rather than taking their dot product, the two embedding matrices can have different sizes (i.e., different numbers of latent factors). fastai has a function `get_emb_sz` that returns recommended sizes for embedding matrices for your data, based on a heuristic that fast.ai has found tends to work well in practice:

Proposed:
> Since we'll be concatenating the embeddings, rather than taking their dot product, the two embedding matrices can have different sizes (i.e., different numbers of latent factors)

---

- [ ] Shorten Ch8 Q31 context - remove "We will see in the next section..."

Current:
> Better still is to use a tabular model based on user meta data to construct your initial embedding vector. When a user signs up, think about what questions you could ask them that could help you to understand their tastes. Then you can create a model where the dependent variable is a user's embedding vector, and the independent variables are the results of the questions that you ask them, along with their signup metadata. We will see in the next section how to create these kinds of tabular models. (You may have noticed that when you sign up for services such as Pandora and Netflix, they tend to ask you a few questions about what genres of movie or music you like; this is how they come up with your initial collaborative filtering recommendations.)

Proposed:
> Better still is to use a tabular model based on user meta data to construct your initial embedding vector. When a user signs up, think about what questions you could ask them that could help you to understand their tastes. Then you can create a model where the dependent variable is a user's embedding vector, and the independent variables are the results of the questions that you ask them, along with their signup metadata

---

- [ ] Shorten Ch9 Q11 - remove "The Kaggle training data end..."

Current:
> If you look at the date range represented in the test set, you will discover that it covers a six-month period from May 2012, which is later in time than any date in the training set. This is a good design, because the competition sponsor will want to ensure that a model is able to predict the future. But it means that if we are going to have a useful validation set, we also want the validation set to be later in time than the training set. The Kaggle training data ends in April 2012, so we will define a narrower training dataset which consists only of the Kaggle training data from before November 2011, and we'll define a validation set consisting of data from after November 2011.

Proposed:
> If you look at the date range represented in the test set, you will discover that it covers a six-month period from May 2012, which is later in time than any date in the training set. This is a good design, because the competition sponsor will want to ensure that a model is able to predict the future. But it means that if we are going to have a useful validation set, we also want the validation set to be later in time than the training set

---

- [ ] Add second option to contexts for Ch10 Q11

Current:

```
{
            "chapter": 10,
            "question_number": 11,
            "question_text": "\"What is xxbos?\"",
            "gold_standard_answer": "\"This is a special token added by fastai that indicated the beginning of the text.\"",
            "answer_context": [
                {
                    "answer_component": "\"This is a special token added by fastai that indicated the beginning of the text.\"", 
                    "scoring_type": "simple",
                    "context": [
                        "- `xxbos`:: Indicates the beginning of a text (here, a review)\n- `xxmaj`:: Indicates the next word begins with a capital (since we lowercased everything)\n- `xxunk`:: Indicates the word is unknown"
                    ],
                    "explicit_context": "true",
                    "extraneous_answer": "false"
                }
            ],
            "question_context": []
        }
```

Proposed:

```
{
            "chapter": 10,
            "question_number": 11,
            "question_text": "\"What is xxbos?\"",
            "gold_standard_answer": "\"This is a special token added by fastai that indicated the beginning of the text.\"",
            "answer_context": [
                {
                    "answer_component": "\"This is a special token added by fastai that indicated the beginning of the text.\"", 
                    "scoring_type": "simple",
                    "context": [
                        "- `xxbos`:: Indicates the beginning of a text (here, a review)\n- `xxmaj`:: Indicates the next word begins with a capital (since we lowercased everything)\n- `xxunk`:: Indicates the word is unknown",
                        "`xxbos`, is a special token that indicates the start of a new text",
                    ],
                    "explicit_context": "true",
                    "extraneous_answer": "false"
                }
            ],
            "question_context": []
        }
```
---

**Comment by vishalbakshi** [comment]
Found this one while working on PassageSpanner:

- [ ] Replace lowercase `t` with upper `T` in Chapter 4 Question 20

Current:

```
{
                "answer_component": "The model therefore cannot learn from the gradients equal to zero, and the model\u2019s weights will not update and will not train",
                "scoring_type": "simple",
                "context": [
                    "this means it is not useful to use accuracy as a loss function\u2014if we do, most of the time our gradients will actually be 0, and the model will not be able to learn from that number",
                    "In mathematical terms, accuracy is a function that is constant almost everywhere (except at the threshold, 0.5), so its derivative is nil almost everywhere (and infinity at the threshold). This then gives gradients that are 0 or infinite, which are useless for updating the model."
                ],
                "explicit_context": "true",
                "extraneous_answer": "false"
            }
```

Proposed:

```
{
                "answer_component": "The model therefore cannot learn from the gradients equal to zero, and the model\u2019s weights will not update and will not train",
                "scoring_type": "simple",
                "context": [
                    "This means it is not useful to use accuracy as a loss function\u2014if we do, most of the time our gradients will actually be 0, and the model will not be able to learn from that number",
                    "In mathematical terms, accuracy is a function that is constant almost everywhere (except at the threshold, 0.5), so its derivative is nil almost everywhere (and infinity at the threshold). This then gives gradients that are 0 or infinite, which are useless for updating the model."
                ],
                "explicit_context": "true",
                "extraneous_answer": "false"
            }
```

**Comment by vishalbakshi** [comment]
Found this while evaluating LLM responses based on answerai-colbert retrieved passages:

Ch1Q33: remove unnecessary sentence in the context:

Current: 

> "if you ensure that you really understand what test and validation sets are and why they're important, then you'll avoid the single biggest source of failures we've seen when organizations decide to use AI. For instance, if you're considering bringing in an external vendor or service, make sure that you hold out some test data that the vendor *never gets to see*. Then *you* check their model on your test data, using a metric that *you* choose based on what actually matters to you in practice, and *you* decide what level of performance is adequate. (It's also a good idea for you to try out some simple baseline yourself, so you know what a really simple model can achieve. <mark>Often it'll turn out that your simple model performs just as well as one produced by an external \"expert\"!)</mark>"

Proposed:

> "if you ensure that you really understand what test and validation sets are and why they're important, then you'll avoid the single biggest source of failures we've seen when organizations decide to use AI. For instance, if you're considering bringing in an external vendor or service, make sure that you hold out some test data that the vendor *never gets to see*. Then *you* check their model on your test data, using a metric that *you* choose based on what actually matters to you in practice, and *you* decide what level of performance is adequate. (It's also a good idea for you to try out some simple baseline yourself, so you know what a really simple model can achieve"

**Comment by vishalbakshi** [comment]
Found this while evaluating LLM responses based on answerai-colbert retrieved passages:

Ch4Q20: Combined first and fourth answer components as they reference the same context.

```
{
                    "answer_component": "A loss function needs to change as the weights are being adjusted", 
                    "scoring_type": "simple",
                    "context": [
                        "we need a loss function which, when our weights result in slightly better predictions, gives us a slightly better loss"
                    ],
                    "explicit_context": "true",
                    "extraneous_answer": "false"
                }
```

```
{
                    "answer_component": "A good loss function gives a slightly better loss when the model gives slightly better predictions", 
                    "scoring_type": "simple",
                    "context": [
                        "Instead, we need a loss function which, when our weights result in slightly better predictions, gives us a slightly better loss"
                    ],
                    "explicit_context": "true",
                    "extraneous_answer": "false"
                }
```
